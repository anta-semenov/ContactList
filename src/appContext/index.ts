import { createContext } from 'react'

export const AppContext = createContext({
  activeContactIndex: 0,
})
