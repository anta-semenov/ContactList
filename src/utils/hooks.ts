import * as React from 'react'
import { Animated, LayoutChangeEvent } from 'react-native'
import { LayoutState, Point } from 'types'

export const useLayout = () => {
  const [layout, setLayout] = React.useState<LayoutState>({ topX: 0, topY: 0, width: 0, height: 0 })
  const center = React.useMemo(() => ({ x: layout.topX + layout.width * 0.5, y: layout.topY + layout.height * 0.5 }), [layout])
  return React.useMemo(() => ({
    layout,
    onLayout: (event: LayoutChangeEvent) => setLayout({
      topX: event.nativeEvent.layout.x,
      topY: event.nativeEvent.layout.y,
      width: event.nativeEvent.layout.width,
      height: event.nativeEvent.layout.height,
    }),
    getCenter: (): Point => center,
  }), [layout])
}

export interface TimingAnimationConfig extends Omit<Animated.TimingAnimationConfig, 'useNativeDriver'> {
  readonly useNativeDriver?: boolean
}

export interface SpringAnimationConfig extends Omit<Animated.SpringAnimationConfig, 'useNativeDriver'> {
  readonly useNativeDriver?: boolean
}

export interface UseAnimationHook {
  readonly value: Animated.Value
  readonly startTiming: (config: TimingAnimationConfig) => Promise<boolean>
  readonly startSpring: (config: SpringAnimationConfig) => Promise<boolean>
  readonly addListener: (callback: AnimationHookListener) => void
}

type AnimationHookListener = (value: number) => void

export const useAnimation = (initialAnimationValue: number, onMountAnimationConfig?: SpringAnimationConfig | TimingAnimationConfig, onMountAnimationCallback?: () => void): UseAnimationHook => {
  const [animationValue] = React.useState(new Animated.Value(initialAnimationValue))
  const isMounted = React.useRef(false)
  const animations = React.useRef<Animated.CompositeAnimation[]>([])
  const listeners = React.useRef<string[]>([])
  React.useEffect(
    () => {
      isMounted.current = true
      if (onMountAnimationConfig != null) {
        if ((onMountAnimationConfig as Animated.TimingAnimationConfig).duration) {
          const animation = Animated.timing(animationValue, { useNativeDriver: true , ...onMountAnimationConfig })
          animation.start(onMountAnimationCallback)
          animations.current = [...animations.current, animation]
        } else {
          const animation = Animated.spring(animationValue, { useNativeDriver: true , ...onMountAnimationConfig as SpringAnimationConfig })
          animation.start(onMountAnimationCallback)
          animations.current = [...animations.current, animation]
        }
      }
      return () => {
        isMounted.current = false
        animations.current.forEach((animation) => animation.stop())
        animations.current = []
        listeners.current.forEach((id) => animationValue.removeListener(id))
        listeners.current = []
      }
    },
    [],
  )

  const addListener = (callback: AnimationHookListener) => {
    const listener = animationValue.addListener(({ value }) => callback(value))
    listeners.current = [...listeners.current, listener]
  }

  return {
    startTiming: (config: TimingAnimationConfig) => new Promise((resolve) => {
      if (isMounted.current === false) {
        resolve(false)
        return
      }
      const animation = Animated.timing(animationValue, { useNativeDriver: true , ...config })
      animation.start(({ finished }) => resolve(finished))
      animations.current = [...animations.current, animation]
    }),
    startSpring: (config: SpringAnimationConfig) => new Promise((resolve) => {
      if (isMounted.current === false) {
        resolve(false)
        return
      }
      const animation = Animated.spring(animationValue, { useNativeDriver: true , ...config })
      animation.start(({ finished }) => resolve(finished))
      animations.current = [...animations.current, animation]
    }),
    value: animationValue,
    addListener,
  }
}

export const useIsFirstMount = () => {
  const isMountRef = React.useRef(true);
  React.useEffect(() => {
    isMountRef.current = false;
  }, []);
  return isMountRef.current;
}
