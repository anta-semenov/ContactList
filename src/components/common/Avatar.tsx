import * as React from 'react'
import { ImageRequireSource, Animated, StyleSheet, View, Image } from 'react-native'
import { avatarSize } from 'constants/'
import { useAnimation, useIsFirstMount } from 'utils'
import { AppContext } from 'appContext'

interface AvatarProps {
  image: ImageRequireSource
  index: number
}

export const Avatar: React.FunctionComponent<AvatarProps> = (props): React.ReactElement => {
  const { activeContactIndex } = React.useContext(AppContext)

  const activeAnimation = useAnimation(props.index === activeContactIndex ? 1 : 0)

  const activeStyle = {
    opacity: activeAnimation.value,
  }

  const isFirstMount = useIsFirstMount()

  React.useEffect(() => {
    if (!isFirstMount) {
      activeAnimation.startSpring({ toValue: activeContactIndex === props.index ? 1 : 0 })
    }
  }, [activeContactIndex === props.index])

  return React.useMemo(() => {
    return (
      <View style={ styles.container }>
        <Animated.View style={ [styles.ring, activeStyle] }/>
        <Image source={ props.image } style={ styles.image } testID='AvatarImage'/>
      </View>
    )
  }, [props.image, activeContactIndex !== props.index])
}

const styles = StyleSheet.create({
  image: {
    width: avatarSize - 20,
    height: avatarSize - 20,
    resizeMode: 'contain',
  },
  container: {
    width: avatarSize,
    height: avatarSize,
    justifyContent: 'center',
    alignItems: 'center',
  },
  ring: {
    position: 'absolute',
    width: avatarSize - 10,
    height: avatarSize - 10,
    borderRadius: (avatarSize - 10) / 2,
    borderWidth: 4,
    borderColor: '#83b8e9',
  },
})
