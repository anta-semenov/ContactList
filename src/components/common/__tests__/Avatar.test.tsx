import * as React from 'react'
import { render } from '@testing-library/react-native'
import { Avatar } from '../Avatar'
import { AppContext } from 'appContext'
import * as hooks from 'utils/hooks'

jest.mock('utils/hooks', () => {
  const startAnimationMock = jest.fn(() => Promise.resolve(true))
  return {
    ...jest.requireActual('utils/hooks') as any,
    startAnimationMock,
    useAnimation: jest.fn((value: number) => ({
      startSpring: startAnimationMock,
      value,
    })),
  }
})



describe('Avatar component', () => {
  it('Should render comoponent image', async () => {
    expect.hasAssertions()

    const { findByTestId } = render(
      <AppContext.Provider value={ { activeContactIndex: 3 } }>
        <Avatar image={ 2 } index={ 1 } />
      </AppContext.Provider>
    )

    const image = await findByTestId('AvatarImage')

    expect(image).toBeTruthy()
  })

  it('Should set active animation vakue to 0 on initial render when avatar is not active', async () => {
    expect.hasAssertions();
    (hooks.useAnimation as jest.MockedFunction<any>).mockClear();
    (hooks as any).startAnimationMock.mockClear()

    render(
      <AppContext.Provider value={ { activeContactIndex: 3 } }>
        <Avatar image={ 2 } index={ 1 } />
      </AppContext.Provider>
    )

    expect((hooks.useAnimation as jest.MockedFunction<any>).mock.calls[0][0]).toBe(0)
    expect((hooks as any).startAnimationMock.mock.calls.length).toBe(0)
  })

  it('Should set active animation vakue to 1 on initial render when avatar is active', async () => {
    expect.hasAssertions();
    (hooks.useAnimation as jest.MockedFunction<any>).mockClear();
    (hooks as any).startAnimationMock.mockClear()

    render(
      <AppContext.Provider value={ { activeContactIndex: 3 } }>
        <Avatar image={ 2 } index={ 3 } />
      </AppContext.Provider>
    )


    expect((hooks.useAnimation as jest.MockedFunction<any>).mock.calls[0][0]).toBe(1)
    expect((hooks as any).startAnimationMock.mock.calls.length).toBe(0)
  })

  it('Should animate when Avatar becomes active', async () => {
    expect.hasAssertions();
    (hooks.useAnimation as jest.MockedFunction<any>).mockClear();
    (hooks as any).startAnimationMock.mockClear()

     const { update } = render(
      <AppContext.Provider value={ { activeContactIndex: 3 } }>
        <Avatar image={ 2 } index={ 1 } />
      </AppContext.Provider>
    )

    update(
      <AppContext.Provider value={ { activeContactIndex: 1 } }>
        <Avatar image={ 2 } index={ 1 } />
      </AppContext.Provider>
    )


    expect((hooks as any).startAnimationMock.mock.calls.length).toBe(1)
    expect((hooks as any).startAnimationMock.mock.calls[0]).toMatchSnapshot()
  })
})
