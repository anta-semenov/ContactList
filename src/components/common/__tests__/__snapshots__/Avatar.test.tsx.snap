// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar component Should animate when Avatar becomes active 1`] = `
Array [
  Object {
    "toValue": 1,
  },
]
`;
